aws/resource_aws_s3_bucket_inventory.go

```diff
@@ -0,0 +1,445 @@
+		inventoryConfiguration.Schedule = &s3.InventorySchedule{
+			Frequency: aws.String(scheduleMap["frequency"].(string)),
+		}
+	}
+
+	if v, ok := d.GetOk("filter"); ok {
+		filterList := v.([]interface{})
+		filterMap := filterList[0].(map[string]interface{})
+		inventoryConfiguration.Filter = expandS3InventoryFilter(filterMap)
+	}
+
+	if v, ok := d.GetOk("destination"); ok {
+		destinationList := v.([]interface{})
+		destinationMap := destinationList[0].(map[string]interface{})
+		bucketList := destinationMap["bucket"].([]interface{})
+		bucketMap := bucketList[0].(map[string]interface{})
+
+		inventoryConfiguration.Destination = &s3.InventoryDestination{
+			S3BucketDestination: expandS3InventoryS3BucketDestination(bucketMap),
+		}
+	}
+
+	input := &s3.PutBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+		InventoryConfiguration: inventoryConfiguration,
+	}
+
+	log.Printf("[DEBUG] Putting S3 bucket inventory configuration: %s", input)
+	err := resource.Retry(1*time.Minute, func() *resource.RetryError {
+		_, err := conn.PutBucketInventoryConfiguration(input)
+		if err != nil {
+			if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") {
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("Error putting S3 bucket inventory configuration: %s", err)
+	}
+
+	d.SetId(fmt.Sprintf("%s:%s", bucket, name))
+
+	return resourceAwsS3BucketInventoryRead(d, meta)
+}
+
+func resourceAwsS3BucketInventoryDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket, name, err := resourceAwsS3BucketInventoryParseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	input := &s3.DeleteBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Deleting S3 bucket inventory configuration: %s", input)
+	_, err = conn.DeleteBucketInventoryConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
+			return nil
+		}
+		return fmt.Errorf("Error deleting S3 bucket inventory configuration: %s", err)
+	}
+
+	return nil
+}
+
+func resourceAwsS3BucketInventoryRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket, name, err := resourceAwsS3BucketInventoryParseID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	d.Set("bucket", bucket)
+	d.Set("name", name)
+
+	input := &s3.GetBucketInventoryConfigurationInput{
+		Bucket: aws.String(bucket),
+		Id:     aws.String(name),
+	}
+
+	log.Printf("[DEBUG] Reading S3 bucket inventory configuration: %s", input)
+	output, err := conn.GetBucketInventoryConfiguration(input)
+	if err != nil {
+		if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
```

Comment:
Using [`d.IsNewResource()`](https://godoc.org/github.com/hashicorp/terraform/helper/schema#ResourceData.IsNewResource) as a conditional for guidance, we should setup the resource to allow retries when recently created due to S3 eventual consistency. e.g.

```go
var output s3.GetBucketInventoryConfigurationOutput
err := resource.Retry(1*time.Minute, func() resource.RetryError {
  var err error
  output, err = conn.GetBucketInventoryConfiguration(input)
  if err != nil {
    if isAWSErr(err, s3.ErrCodeNoSuchBucket, "") || isAWSErr(err, "NoSuchConfiguration", "The specified configuration does not exist.") {
      if d.IsNewResource() {
        return resource.RetryableError(err)
      }
      // output should be nil in this case and caught later (or you can duplicate the NoSuchBucket/Configuration conditional in resource.Retry err check to remove resource from state)
      return nil
    }
    return resource.NonRetryableError(err)
  }
  return nil
})
if err != nil {
  return fmt.Errorf("error reading S3 Bucket Inventory Configuration: %s", err)
}

if output == nil || output.InventoryConfiguration == nil {
  log.Printf("[WARN] %s S3 bucket inventory configuration not found, removing from state.", d.Id())
  d.SetId("")
  return nil
}
```